crates/storage/state_handler/src/lib.rs

```diff
@@ -0,0 +1,82 @@
+//! A state handler that can be used to store and retrieve state.
+mod snapshot_state;
+
+use crate::snapshot_state::{Snapshot, SnapshotState};
+use allocator_api2::alloc::Global;
+use allocator_api2::vec::Vec;
+use ixc_message_api::AccountID;
+
+/// A store that can be used to store and retrieve state.
+pub trait Store {
+    /// Get the value for the given key.
+    fn get(&self, key: &Vec<u8>) -> Option<Vec<u8>>;
+    /// Commit the state changes.
+    fn commit(&mut self) -> Result<(), ()>;
+}
+
+/// StateHandler is a cache-based state handler that can be used to store and retrieve state.
+pub struct StateHandler<S: Store> {
+    snapshot_state: SnapshotState<S>,
+    /// Checkpoints are used to revert state changes.
+    /// checkpoints are used to follow the call stack of a transaction
+    /// and revert the state changes when the transaction is rolled back.
+    checkpoints: Vec<Snapshot>,
+}
+
+impl<S: Store> StateHandler<S> {
+    /// Create a new state handler with the given store.
+    pub fn new(store: S) -> Self {
+        Self {
+            snapshot_state: SnapshotState::new(store),
+            checkpoints: Vec::with_capacity_in(200, Global),
```

Comment:
discussed in the call: would be better as a linkedlist to avoid trashing of memory